server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/WeightFunction.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import org.elasticsearch.cluster.ClusterInfo;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.allocation.WriteLoadForecaster;
+import org.elasticsearch.index.shard.ShardId;
+
+/**
+ * This class is the primary weight function used to create balanced over nodes and shards in the cluster.
+ * Currently this function has 3 properties:
+ * <ul>
+ * <li><code>index balance</code> - balance property over shards per index</li>
+ * <li><code>shard balance</code> - balance property over shards per cluster</li>
+ * </ul>
+ * <p>
+ * Each of these properties are expressed as factor such that the properties factor defines the relative
+ * importance of the property for the weight function. For example if the weight function should calculate
+ * the weights only based on a global (shard) balance the index balance can be set to {@code 0.0} and will
+ * in turn have no effect on the distribution.
+ * </p>
+ * The weight per index is calculated based on the following formula:
+ * <ul>
+ * <li>
+ * <code>weight<sub>index</sub>(node, index) = indexBalance * (node.numShards(index) - avgShardsPerNode(index))</code>
+ * </li>
+ * <li>
+ * <code>weight<sub>node</sub>(node, index) = shardBalance * (node.numShards() - avgShardsPerNode)</code>
+ * </li>
+ * </ul>
+ * <code>weight(node, index) = weight<sub>index</sub>(node, index) + weight<sub>node</sub>(node, index)</code>
+ */
+public class WeightFunction {
```

Comment:
This is basically just moved out of `BalancedShardsAllocator`, with minor changes to not pass `Balancer` and `ModelNode`.